Cache prefetching

wild-boomerang opened the pull request:
Two cache prefetching algorithms have been implemented: next line and wrong path.

wild-boomerang committed:
added prefetching

wild-boomerang committed:
added command line args for prefetching

wild-boomerang committed:
wrong path release

wrong path prefetch method was implemented; also was added command line arguments for prefetch method selection; some small code refactoring

wild-boomerang committed:
wrong path release

wrong path prefetch method was implemented; also was added command line arguments for prefetch method selection; some small code refactoring

wild-boomerang committed:
small fix

pavelkryukov commented on simulator/infra/cache/cache_tag_array.cpp:
```diff
@@ -225,7 +225,7 @@ std::pair<bool, int32> SimpleCacheTagArray::read( Addr addr)
     {
         uint32 num_set = set( addr);
-        replacement_module->touch( num_set, way);
+        replacement_module->touch( num_set, way); // for substitution policy (updates the current address)
```
In CS, we use term "replacement policy" for this.

pavelkryukov commented on simulator/infra/ports/module.cpp:
```diff
@@ -48,6 +48,16 @@ void Module::enable_logging_impl( const std::unordered_set<std::string>& names)
+    this->prefetch_method = prefetch_method;
+    for ( const auto& c : children)
+        c->enable_prefetch(key, fetchahead_size, prefetch_method);
+}
```
Now we keep `config` variables locally to modules.
Although I agree there are architectural disadvantages of this, I propose to be consistent.

pavelkryukov commented on simulator/infra/ports/module.cpp:
```diff
@@ -48,6 +48,16 @@ void Module::enable_logging_impl( const std::unordered_set<std::string>& names)
+{
+    prefetch_enabled = key;
+    this->fetchahead_size = fetchahead_size;
+    this->prefetch_method = prefetch_method;
```
We do not use `this` once it is not needed (see https://github.com/MIPT-ILab/mipt-mips/wiki/Cpp-guidelines#this).

Also, here parameters and members have the same name — warned by Visual Studio.

pavelkryukov commented on simulator/infra/ports/module.h:
```diff
@@ -42,6 +42,11 @@ class Module : public Log
+    void enable_prefetch(bool key, uint32 fetchahead_size, std::string prefetch_method); // enable-prefetch function initialization
+    bool prefetch_enabled = false; // boolean value that represents that prefetch is enabled/disabled
+    uint32 fetchahead_size; // value of fetchahead distance size
+    std::string prefetch_method; // value of prefetch method
```
The problem is that _all_ modules have knowledge regarding the prefetching. That's why we are trying to keep features locally.

pavelkryukov commented on simulator/modules/core/perf_sim.cpp:
```diff
@@ -13,6 +13,12 @@
 namespace config {
     static const AliasedValue<std::string> units_to_log = { "l", "logs", "nothing", "print logs for modules"};
     static const Switch topology_dump = { "tdump", "module topology dump into topology.json" };
+    // command line flag argument for enable/disable prefetch
```
This comment does not bring any value compared to code. `AliasedSwitch` is a command line flag argument, and its string description must describe it enough.

pavelkryukov commented on simulator/modules/core/perf_sim.cpp:
```diff
@@ -13,6 +13,12 @@
     static const Switch topology_dump = { "tdump", "module topology dump into topology.json" };
+    // command line flag argument for enable/disable prefetch
+    static const AliasedSwitch prefetch = { "p", "prefetch", "cache prefetching" };
+    // command line argument for fetchahead distance size
```
Same comment.

pavelkryukov commented on simulator/modules/core/perf_sim.cpp:
```diff
@@ -13,6 +13,12 @@
+    static const AliasedSwitch prefetch = { "p", "prefetch", "cache prefetching" };
+    // command line argument for fetchahead distance size
+    static const Value<uint32> fetchahead_distance = { "fetchahead-size", 32, "fetchahead distance size"};
+    // command line argument for prefetch type
```
Same comment.

pavelkryukov commented on simulator/modules/core/perf_sim.cpp:
```diff
@@ -110,6 +117,10 @@ void PerfSim<ISA>::dump_statistics() const
+    // info to output about prefetch
+    std::string is_prefetch = (config::prefetch) ? "yes" : "no";
+    uint32 fetchahead_distance = config::fetchahead_distance;
+    std::string prefetch_type = (config::prefetch_method == "wrong_path") ? "wrong path" : "next line";
```
I don't think we need all options dumped, since they are part of command line.

pavelkryukov commented on simulator/modules/fetch/fetch.cpp:
```diff
@@ -67,15 +67,27 @@ Target Fetch<FuncInstr>::get_target( Cycle cycle)
     const Target branch_target   = rp_target->is_ready( cycle) ? rp_target->read( cycle) : Target();
 
+    if (prefetch_enabled && prefetch_method == "wrong_path")
+    {
```
Our code guideline is to use as less brackets as possible: https://github.com/MIPT-ILab/mipt-mips/wiki/Code-style-guidelines#pythonesque-statements

pavelkryukov commented on simulator/modules/fetch/fetch.cpp:
```diff
@@ -67,15 +67,27 @@ Target Fetch<FuncInstr>::get_target( Cycle cycle)
+        if (prefetch_enabled && prefetch_method == "wrong_path") // prefetch wrong path if prefetch enabled and defined
+        {
+            is_wrong_path = true;
+            tags->write( flushed_target_from_decode.address);
```
If I understand correctly. this is not precise, as cache miss latency is not accounted here.
Also, what would happen if a line is already presented in the instruction cache?

wild-boomerang replied:
I added a check if the line is present in the cache or not. I think writing to cache is in the right place because the prefetch method tries to prevent cache misses. But if I am wrong about this, please explain to me.

pavelkryukov commented on simulator/modules/fetch/fetch.cpp:
```diff
@@ -188,6 +200,30 @@ void Fetch<FuncInstr>::clock( Cycle cycle)
     wp_datapath->write( std::move( instr), cycle);
+
+    if ( prefetch_enabled && !is_wrong_path)  // prefetch next line if enabled and wrong path isn't used
+    {
```
Our code guideline is to use as less brackets as possible: https://github.com/MIPT-ILab/mipt-mips/wiki/Code-style-guidelines#pythonesque-statements

pavelkryukov commented on simulator/modules/fetch/fetch.cpp:
```diff
@@ -188,6 +200,30 @@ void Fetch<FuncInstr>::clock( Cycle cycle)
+template<typename FuncInstr>
+void Fetch<FuncInstr>::prefetch_next_line(Addr requested_addr)
+{
+    uint32 line_size = uint32(config::instruction_cache_line_size); // line size
```
Use type once: `uint32 line_size(config::instruction_cache_line_size)`

pavelkryukov commented on simulator/modules/fetch/fetch.cpp:
```diff
@@ -188,6 +200,30 @@ void Fetch<FuncInstr>::clock( Cycle cycle)
+
+   /* if the fetchahead is reached and there is no line in the cache, write the line to the cache */
+   if ( offset >= (line_size - fetchahead_distance) && !tags->lookup( next_line_addr))
+   {
```
Our code guideline is to use as less brackets as possible: https://github.com/MIPT-ILab/mipt-mips/wiki/Code-style-guidelines#pythonesque-statements

pavelkryukov reviewed (commented):
After fixing comments, you'll likely said there is a code coverage regression.
Do you have ideas how can we test the new behavior?

pavelkryukov commented on simulator/modules/core/perf_sim.cpp:
```diff
@@ -13,6 +13,12 @@
+    // command line argument for fetchahead distance size
+    static const Value<uint32> fetchahead_distance = { "fetchahead-size", 32, "fetchahead distance size"};
+    // command line argument for prefetch type
+    static const Value<std::string> prefetch_method = {"prefetch-method", "wrong_path", "Type of prefetching method"};
```
Can we use "prefetch_method" variable to enable/disable prefetches as well by having a third option "no-prefetch"?
Also, please specify that is a Instruction prefetch.
We are likely to have data prefetches once there is data cache.

wild-boomerang replied:
Is it worth deleting the "prefetch" variable in this case?

pavelkryukov replied:
Yes

wild-boomerang committed:
Transfer prefetch config variables to fetch module

wild-boomerang committed:
Remove prefetch_enabled

prefetch_method variable now have the third option "no-prefetch";
Add PrefetchMethodException struct;
Remove enable-prefetch method because of its useless.

pavelkryukov committed:
Merge branch 'master' into cache_prefetching

wild-boomerang commented:
What should I do now?

pavelkryukov committed:
Merge branch 'master' into cache_prefetching

wild-boomerang commented:
Do I need to write tests to increase code coverage? And after that you can merge into master branch?

pavelkryukov commented:
> Do I need to write tests to increase code coverage? 

This is welcome, but not mandatory.
There is no good infrastructure to test modules separately, so I'll merge it to carry on later.
Could you please introduce yourself so we can mention you in LICENSE?

wild-boomerang commented:
Ok, I understood. 
My name is Vladislav Zhuravski.

pavelkryukov commented:
Thank you for your contribution!

wild-boomerang commented:
Thank you for such a good opportunity to contribute!
